[AOTAutograd] compare with stride hints

shunting314 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #103342

We previously compare FakeTensor's strides with real tensor's strides. This cause dynamic dimension of FakeTensor being specialized to static int. This may cause a graph specialized for one shape being used by another shape which is wrong.

Use stride hints for the comparison instead.



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @aakhundov

shunting314 committed:
[AOTAutograd] compare with stride hints

[ghstack-poisoned]

shunting314 committed:
Update on "[AOTAutograd] compare with stride hints"


We previously compare FakeTensor's strides with real tensor's strides. This cause dynamic dimension of FakeTensor being specialized to static int. This may cause a graph specialized for one shape being used by another shape which is wrong.

Use stride hints for the comparison instead.



[ghstack-poisoned]

shunting314 commented on test/inductor/test_layout_optim.py:
```diff
@@ -238,6 +238,28 @@ def get_example_inputs(self):
 
         self.verify_accuracy_for_infer(Model)
 
+    def test_dyanmic_shape_specialization(self):
```
The test fail with the similar symptom as internal test https://www.internalfb.com/tasks?t=155292157 before my fix and pass after.

malfet commented on torch/_functorch/aot_autograd.py:
```diff
@@ -662,6 +663,51 @@ def from_fun(t):
+        new_strides[i] = next_stride
+        next_stride = next_stride * sizes[i]
+
+    # convert sympy exprssion to SymInt
```
```suggestion
    # convert sympy expression to SymInt
```

shunting314 committed:
Update on "[AOTAutograd] compare with stride hints"


We previously compare FakeTensor's strides with real tensor's strides. This cause dynamic dimension of FakeTensor being specialized to static int. This may cause a graph specialized for one shape being used by another shape which is wrong.

Use stride hints for the comparison instead.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 aakhundov

[ghstack-poisoned]

shunting314 committed:
Update on "[AOTAutograd] compare with stride hints"


We previously compare FakeTensor's strides with real tensor's strides. This cause dynamic dimension of FakeTensor being specialized to static int. This may cause a graph specialized for one shape being used by another shape which is wrong.

Use stride hints for the comparison instead.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 aakhundov

[ghstack-poisoned]

shunting314 commented:
@pytorchbot merge

shunting314 committed:
Update on "[AOTAutograd] compare with stride hints"


We previously compare FakeTensor's strides with real tensor's strides. This cause dynamic dimension of FakeTensor being specialized to static int. This may cause a graph specialized for one shape being used by another shape which is wrong.

Use stride hints for the comparison instead.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 aakhundov

[ghstack-poisoned]

shunting314 committed:
Update on "[AOTAutograd] compare with stride hints"


We previously compare FakeTensor's strides with real tensor's strides. This cause dynamic dimension of FakeTensor being specialized to static int. This may cause a graph specialized for one shape being used by another shape which is wrong.

Use stride hints for the comparison instead.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 aakhundov

[ghstack-poisoned]

shunting314 commented on torch/_functorch/aot_autograd.py:
```diff
@@ -3001,7 +3011,26 @@ def call_compiled_backward():
+                        # cause dynamic dimensions in ph_arg being specialized to static
+                        # value. Using the hints to avoid that.
+                        if _get_hints(ph_arg.stride()) != real_arg.stride():
+                            # Note that here we use the stride of the real tensor to
```
CC @ezyang since I don't want to sneak in this change without your attention. I ultimately have to fallback to this 'bad' solution I initially implemented.

The other 'better' solution that decide the stride order from the real tensor and apply this order to the FakeTensor does not always work. Model like mixnet_l indeed uses some tensors with non 'dense' layout. Check the code comments here for more details. Apply a stride order to a FakeTensor will always generate a tensor with 'dense' layout.

Right now since we disable layout optimization for dynamic shape (for the other perf reason lol), it's safe to apply the stride from the real tensor directly to the fake tensor.

I'll be happy to discuss more but would like to get this fix in first to unblock.

ezyang replied:
You can still handle the non-dense case with permutation. But the most important thing is to do the tests *in forwards*, because we have no way of adding guards if they show up in backwards right now.

shunting314 commented:
@pytorchbot merge -f "verified the failed test fail on base commit as well"
